source/_posts/2019-07-15-complete-flows-partial-models.markdown

```diff
@@ -0,0 +1,166 @@
+---
+layout: post
+author: rafa
+title: "(Complete, Flows) -> (Partial) -> (Models)"
+date: 2019-07-15 19:55:56 +0200
+comments: true
+categories: 
+---
+
+Most apps these days have a sequence of screens that gather information from the user, like a registration flow, a form of some kind. The data from each step is typically combined into a single data structure.
+For example, let's say we want the name, age, and the password to authenticate the user.
+<!--more-->
+One way to model it is by using the following data structure:
+
+```swift
+struct FormData {
+    let name: String
+    let age: Int
+    let password: String
+}
+```
+
+One issue we are going to come about is that our model is strict, it needs all the values at once, whereas users will supply each value at a time. First they will type in their name, then their age, and so on.
+Wrapping up the fields in `Optional`, may loosen its strictness.
+
+```swift
+struct FormData {
+    let name: String?
+    let age: Int?
+    let password: String?
+}
+```
+
+Our flow code might look like:
+
+```swift
+func firstStepFinished(with name: String) -> FormData {
+    return FormData(name: name, age: nil, password: nil)
+}
+
+func secondStepFinished(with age: Int, partialFormData: FormData) -> FormData {
+    return FormData(name: partialFormData.name, age: age, password: nil)
+}
+
+func thirdStepFinished(with password: String, partialFormData: FormData) {
+    let formData = FormData(name: partialFormData.name, age: partialFormData.age, password: password)
+    
+    api.performLogin(with: formData)
+}
+```
+
+However, now we need to `guard` against any `nil` values if we want to use them (for example, to make a network request).
```

Comment:
Also super obvious, but.. Why guarding against `nil` values here is bad? :smiling_imp:

Reply:
what do you think would be a good way to tackle this issue?

I'm trying to wrap up the problem with Optionals later on

```
From a domain perspective, that `return` doesn't make any sense. 
...
```

Reply:
maybe @mmatoszko means another option would be to force unwrap, given that would resolve your domain concern. Maybe you can quote https://www.swiftbysundell.com/posts/handling-non-optional-optionals-in-swift and say that this architecture has effectively come up with non-optional optionals at the end of the flow!